docs/tutorials/cooperative_groups_tutorial.rst

```diff
@@ -0,0 +1,312 @@
+.. code-block:: cpp
+
+  /// \brief Summation of `unsigned int val`'s in `thread_group g` using shared memory `x`
+  __device__ unsigned int reduce_sum(thread_group g, unsigned int* x, unsigned int val)
+  {
+      // Rank of this thread in the group
+      const unsigned int group_thread_id = g.thread_rank();
+
+      // We start with half the group size as active threads
+      // Every iteration the number of active threads halves, until we processed all values
+      for(unsigned int i = g.size() / 2; i > 0; i /= 2)
+      {
+          // Store value for this thread in a shared, temporary array
+          x[group_thread_id] = val;
+
+          // Synchronize all threads in the group
+          g.sync();
+
+          // If our thread is still active, sum with its counterpart in the other half
+          if(group_thread_id < i)
+          {
+              val += x[group_thread_id + i];
+          }
+
+          // Synchronize all threads in the group
+          g.sync();
+      }
+
+      // Only the first thread returns a valid value
+      if(g.thread_rank() == 0)
+          return val;
+      else
+          return 0;
+  }
+
+The ``reduce_sum`` device function reused to calculate the block and custom
+partition sum of the input numbers. The kernel has three different sections:
+1. The reduction function variables initialization.
+2. The reduction of thread block and store the results in global memory.
+3. The reduction of custom partition and store the results in global memory.
+
+
+1. The reduction function variables initialization
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+In this code section the shared memory is declared, the thread_block_group and
+custom_partition are defined and the input variable are loaded from global
+memory.
+
+.. code-block:: cpp
+
+  // threadBlockGroup consists of all threads in the block
+  thread_block thread_block_group = this_thread_block();
+
+  // Workspace array in shared memory required for reduction
+  __shared__ unsigned int workspace[2048];
+
+  unsigned int output;
+
+  // Input to reduce
+  const unsigned int input = d_vector[thread_block_group.thread_rank()];
+
+  // ...
+
+  // Every custom_partition group consists of 16 threads
+  thread_block_tile<PartitionSize> custom_partition
+          = tiled_partition<PartitionSize>(thread_block_group);
+
+
+2. The reduction of thread block
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+In this code section the sum is calculate on thread_block_group level, then the
+results are stored in global memory.
+
+.. code-block:: cpp
+
+  // Perform reduction
+  output = reduce_sum(thread_block_group, workspace, input);
+
+  // Only the first thread returns a valid value
+  if(thread_block_group.thread_rank() == 0)
+  {
+    d_block_reduced_vector[0] = output;
+  }
+
+3. The reduction of custom partition
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+In this code section the sum is calculate on thread_block_group level, then the
+results are stored in global memory.
```

Comment:
i get it, that it does mostly the same thing as the last step, but there should be an explainer with the difference

Reply:
Valid point. I hope I fixed this.